### PR TITLE
more flexibility for autocmd control

### PR DESCRIPTION
**Pull request by xiaoshihou514**
Also tried to make functions in init.lua more descriptive in terms of naming

### CODE REVIEW

**Comment by xiaoshihou514** [comment]
I don't know if the design decision is correct :cry:, please correct me if it feels wrong to you

File: lua/guard/util.lua [hunk 1]
```diff
@@ -29,4 +31,104 @@ function util.as_table(t)
...
+     if not client.supports_method('textDocument/formatting') then
+       return
+     end
+     if ft_handler[vim.bo[args.buf].filetype] and ft_handler[vim.bo[args.buf].filetype].format then
+       table.insert(ft_handler[vim.bo[args.buf].filetype].format, 1, 'lsp')
+     else
+       ft_handler(vim.bo[args.buf].filetype):fmt('lsp')
+     end
+
+     local ok, au = pcall(api.nvim_get_autocmds, {
```
Comment (glepnir) [inline lua/guard/util.lua:69]: is there necessary use pcall  ? looks like if no autocmd it just return an empty table not through an error. If i remember correct ? forget the source code of `nvim_get_autocmds`.

File: lua/guard/util.lua [hunk 2]
```diff
@@ -29,4 +31,104 @@ function util.as_table(t)
   return vim.tbl_islist(t) and t or { t }
 end
 
+function util.attach_to(buf)
```
Comment (glepnir) [inline lua/guard/util.lua:34]: these functions are doing the format things. so they are not `util` . in this module looks wired.

File: plugin/guard.lua [hunk 3]
```diff
@@ -1,9 +0,0 @@
-local loaded = false
-if loaded then
-  return
-end
-loaded = true
-
-vim.api.nvim_create_user_command('GuardFmt', function()
```
Comment (glepnir) [inline plugin/guard.lua:7]: if command in there user can use `GuardFmt` to lazyload when they set ` fmt_on_save ` to false.

File: lua/guard/util.lua (hunk 1 above)
Comment (glepnir) [reply lua/guard/util.lua:60]: okay . fair enough

File: plugin/guard.lua (hunk 3 above)
Comment (xiaoshihou514) [reply plugin/guard.lua:7]: Gotcha

File: lua/guard/util.lua [hunk 4]
```diff
@@ -104,6 +103,7 @@ end
 
 function util.enable(opts)
```
Comment (glepnir) [inline lua/guard/util.lua:104]: they are not belong to `util` module. in `event ` module will better.

Comment (xiaoshihou514) [reply lua/guard/util.lua:104]: I am thinking event.lua(for attach, watch and lspattach) and api.lua (for disable and enable)?

Comment (glepnir) [reply lua/guard/util.lua:104]: yup :)

**Comment by xiaoshihou514** [comment]
[This](https://github.com/nvimdev/guard.nvim/blob/ccdd3abbf71660828d6a34eca30057d6d8308901/lua/guard/init.lua#L82-L86) seems to lead to `require("guard.filetype").foo = { "lsp", "lsp", "lsp" }` if one does multiple `LspRestart`? So I changed it to only add to the table when there are no existing formatters.

File: lua/guard/api.lua [hunk 5]
```diff
@@ -0,0 +1,72 @@
...
+    for _, au in ipairs(guard_aus) do
+      api.nvim_del_autocmd(au.id)
+    end
+    return
+  end
+  if #guard_aus == 0 then
+    return
+  end
+  local arg = opts.args
+  local _, bufnr = pcall(tonumber, arg)
```
Comment (glepnir) [inline lua/guard/api.lua:18]: there can  use regex  like `arg:find('%d+') ` is buffer `arg:find('%w+')` is filetype . and if `#args == 0 ` should be current buffer .

Comment (glepnir) [reply lua/guard/api.lua]: I think this not resolved ?

File: lua/guard/api.lua [hunk 6]
```diff
@@ -0,0 +1,70 @@
+local api = vim.api
+local ft_handler = require('guard.filetype')
+local events = require('guard.events')
+local format = require('guard.format')
+
+local function disable(opts)
+  if #opts.fargs == 0 then
+    local current = api.nvim_get_autocmds({ group = 'Guard', event = 'BufWritePre', buffer = api.nvim_get_current_buf() })
```
Comment (glepnir) [inline lua/guard/api.lua:8]: there are duplicate . you can do it on bufferr check

File: lua/guard/api.lua [hunk 7]
```diff
@@ -0,0 +1,55 @@
...
+
+local function disable(opts)
+  local arg = opts.args
+  local bufnr = (#opts.fargs == 0) and api.nvim_get_current_buf() or tonumber(arg)
+  if bufnr then
+    local bufau = api.nvim_get_autocmds({ group = 'Guard', event = 'BufWritePre', buffer = bufnr })
+    if #bufau ~= 0 then
+      api.nvim_del_autocmd(bufau[1].id)
+    end
+  else
```
Comment (glepnir) [inline lua/guard/api.lua:14]: I am wonder do we really need this ?  disable in a filetype ...

**Pull request merged**